### PR TITLE
Linq: add enum HasFlag method support

### LINKED ISSUE #829: NH-3365 - Support for HasFlag method for enums with Flags attribute applied.
**Andrey Kozhyn** created an issue — <time datetime="2013-01-02T15:19:09.924Z" title="2013-01-02T15:19:09.924Z">2nd January 2013, 15:19:09</time>:
> When HasFlag method is called for any enum System.NotSupportedException is thrown for such query:
> 
> session.Query<Entity>().FirstOrDefault(x => x.EnumProperty.HasFlag(TestEnum.Value));
> 
---
**Alexander Zaytsev** added a comment — <time datetime="2013-01-02T15:21:43.223Z" title="2013-01-02T15:21:43.223Z">2nd January 2013, 15:21:43</time>:
> Does EntityFramework or Linq2Sql support this?
> 
---
**Andrey Kozhyn** added a comment — <time datetime="2013-01-02T15:37:20.015Z" title="2013-01-02T15:37:20.015Z">2nd January 2013, 15:37:20</time>:
> Checked for EntityFramework. It doesn't support it.
> 
---
**Andreas Eriksson** added a comment — <time datetime="2015-07-07T20:24:33.313Z" title="2015-07-07T20:24:33.313Z">7th July 2015, 20:24:33</time>:
> Entity Framework has support for it since 6.1 
> http://data.uservoice.com/forums/72025-entity-framework-feature-suggestions/suggestions/3951293-support-enum-hasflag
>

### PR DESCRIPTION
**Pull request by tykovec**
Closes #829

### CODE REVIEW

**Comment by tykovec** [comment]
Works only with enum persisted as number.

File: src/NHibernate/Linq/ExpressionTransformers/RemoveRedundantCast.cs [hunk 1]
```diff
@@ -20,6 +21,7 @@ public class RemoveRedundantCast : IExpressionTransformer<UnaryExpression>
 		public Expression Transform(UnaryExpression expression)
 		{
 			if (expression.Type != typeof(object) &&
+				expression.Type != typeof(Enum) &&
```
Comment (tykovec) [inline src/NHibernate/Linq/ExpressionTransformers/RemoveRedundantCast.cs:24]: This part is fix for exception thrown in SelectChildClassContainedInParent test.

File: src/NHibernate/Linq/Functions/HasFlagGenerator.cs [hunk 2]
```diff
@@ -0,0 +1,34 @@
+﻿using System;
+using System.Collections.ObjectModel;
+using System.Linq.Expressions;
+using System.Reflection;
+using NHibernate.Hql.Ast;
+using NHibernate.Linq.Visitors;
+
+namespace NHibernate.Linq.Functions
+{
+	internal class HasFlagGenerator : BaseHqlGeneratorForMethod, IRuntimeMethodHqlGenerator
```
Comment (hazzik) [inline src/NHibernate/Linq/Functions/HasFlagGenerator.cs:10]: This does not have to be a `IRuntimeMethodHqlGenerator`

This should work:
```csharp
SupportedMethods = new[] { ReflectHelper.FastGetMethod(Enum.HasFlag, default(Enum)) }
```

Comment (hazzik) [reply src/NHibernate/Linq/Functions/HasFlagGenerator.cs:10]: Ok, I tried, it did not work.

Comment (bahusoid) [reply src/NHibernate/Linq/Functions/HasFlagGenerator.cs:10]: It's not static. But should work with 
```C#
SupportedMethods = new[] { ReflectHelper.GetMethodDefinition<Enum>(x => x.HasFlag(default)) };
```

Comment (tykovec) [reply src/NHibernate/Linq/Functions/HasFlagGenerator.cs]: i see, we used it with IRuntimeMethodHqlGenerator, there was some issue in past but with newer version there is no problem i updated my PR

**Review by bahusoid — APPROVED** [review]

**Pull request merged**